File: javascript/exercises.js
import { open } from "node:fs/promises"

export function change(amount) {
  if (!Number.isInteger(amount)) {
    throw new TypeError("Amount must be an integer")
  }
  if (amount < 0) {
    throw new RangeError("Amount cannot be negative")
  }
  let [counts, remaining] = [{}, amount]
  for (const denomination of [25, 10, 5, 1]) {
    counts[denomination] = Math.floor(remaining / denomination)
    remaining %= denomination
  }
  return counts
}

// Write your first then lower case function here
function firstThenLowerCase(sequence, predicate) {
  return sequence.find(predicate)?.toLowerCase();
}

// Write your powers generator here
function* powersGenerator({ ofBase, upTo }) {
  let power = 1;
  while (power <= upTo) {
      yield power;
      power *= ofBase;
  }
}

// Write your say function here

function say(word = "") {
  const words = []; // Array to store the words

  function inner(nextWord) {
    if (nextWord === undefined) {
      return words.join(" "); // Return joined words if no argument is passed
    }
    words.push(nextWord);
    return inner; 
  }

  words.push(word); 

  return inner;

}
// Write your line count function here
const fs = require('fs').promises;

async function meaningfulLineCount(filename) {
    try {
        const data = await fs.readFile(filename, 'utf-8');
        const lines = data.split('\n');

        const count = lines.reduce((acc, line) => {
            const trimmedLine = line.trim();
            if (trimmedLine !== '' && !trimmedLine.startsWith('#')) {
                return acc + 1;
            }
            return acc;
        }, 0);

        return count;
    } catch (error) {
        throw new Error('No such file');  // Propagate the error as a rejection
    }
}
// Write your Quaternion class here
class Quaternion {
  constructor(a, b, c, d) {
    this._a = a;
    this._b = b;
    this._c = c;
    this._d = d;
    Object.freeze(this); // Make the object immutable
  }

  get coefficients() {
    return [this._a, this._b, this._c, this._d];
  }

  // Getter for the conjugate
  get conjugate() {
    return new Quaternion(this._a, -this._b, -this._c, -this._d);
  }

  // Method to add two quaternions
  plus(q) {
    return new Quaternion(this._a + q._a, this._b + q._b, this._c + q._c, this._d + q._d);
  }

  // Method to multiply two quaternions
  times(q) {
    const a = this._a * q._a - this._b * q._b - this._c * q._c - this._d * q._d;
    const b = this._a * q._b + this._b * q._a + this._c * q._d - this._d * q._c;
    const c = this._a * q._c - this._b * q._d + this._c * q._a + this._d * q._b;
    const d = this._a * q._d + this._b * q._c - this._c * q._b + this._d * q._a;
    return new Quaternion(a, b, c, d);
  }

  equals(q) {
    return this._a === q._a && this._b === q._b && this._c === q._c && this._d === q._d;
  }

  toString() {
    let terms = [];

    if (this._a !== 0) terms.push(`${this._a}`);
    if (this._b !== 0) terms.push(`${this._b === 1 ? '' : this._b === -1 ? '-' : this._b}i`);
    if (this._c !== 0) terms.push(`${this._c === 1 ? '' : this._c === -1 ? '-' : this._c}j`);
    if (this._d !== 0) terms.push(`${this._d === 1 ? '' : this._d === -1 ? '-' : this._d}k`);

    return terms.join('+').replace(/\+\-/g, '-'); // Join terms and correct formatting of signs
  }
}

File: python/exercises.py
from dataclasses import dataclass
from collections.abc import Callable


def change(amount: int) -> dict[int, int]:
    if not isinstance(amount, int):
        raise TypeError('Amount must be an integer')
    if amount < 0:
        raise ValueError('Amount cannot be negative')
    counts, remaining = {}, amount
    for denomination in (25, 10, 5, 1):
        counts[denomination], remaining = divmod(remaining, denomination)
    return counts


# Write your first then lower case function here
def first_then_lower_case(sequence, predicate):
    for element in sequence:
        if predicate(element):
            return element.lower()
    return None


# Write your powers generator here
def powers_generator(*, base, limit):
    power = 1
    while power <= limit:
        yield power
        power *= base

def say(word=""):
    words = [] 

    def inner(next_word=None):
        if next_word is None:
            return " ".join(words)
        words.append(next_word)
        return inner  

    words.append(word)

    return inner







def meaningful_line_count(filename):
    count = 0
    try:
        with open(filename, 'r') as file:  # Auto-closes the file
            for line in file:
                stripped_line = line.strip()  # Strip whitespace
                if stripped_line and not stripped_line.startswith('#'):
                    count += 1
    except FileNotFoundError:
        raise FileNotFoundError("No such file") 
    return count

# Write your Quaternion class here


from dataclasses import dataclass

@dataclass(frozen=True)
class Quaternion:
    a: float
    b: float
    c: float
    d: float

    @property
    def coefficients(self):
        return (self.a, self.b, self.c, self.d)

    @property
    def conjugate(self):
        return Quaternion(self.a, -self.b, -self.c, -self.d)

    def __add__(self, other):
        if not isinstance(other, Quaternion):
            return NotImplemented
        return Quaternion(self.a + other.a, self.b + other.b, self.c + other.c, self.d + other.d)

    def __mul__(self, other):
        if not isinstance(other, Quaternion):
            return NotImplemented
        a = self.a * other.a - self.b * other.b - self.c * other.c - self.d * other.d
        b = self.a * other.b + self.b * other.a + self.c * other.d - self.d * other.c
        c = self.a * other.c - self.b * other.d + self.c * other.a + self.d * other.b
        d = self.a * other.d + self.b * other.c - self.c * other.b + self.d * other.a
        return Quaternion(a, b, c, d)

    def __eq__(self, other):
        if not isinstance(other, Quaternion):
            return False
        return (self.a, self.b, self.c, self.d) == (other.a, other.b, other.c, other.d)

    def __str__(self):
        terms = []
        if self.a != 0:
            terms.append(f"{self.a}")

        # Handle 'i' term
        if self.b != 0:
            b_str = f"{self.b}" if abs(self.b) != 1 else ("-" if self.b == -1 else "")
            terms.append(f"{b_str}i")

        # Handle 'j' term
        if self.c != 0:
            c_str = f"{self.c}" if abs(self.c) != 1 else ("-" if self.c == -1 else "")
            terms.append(f"{c_str}j")

        # Handle 'k' term
        if self.d != 0:
            d_str = f"{self.d}" if abs(self.d) != 1 else ("-" if self.d == -1 else "")
            terms.append(f"{d_str}k")

        # Handle the case where all terms are zero
        if not terms:
            return "0"

        # Join the terms, ensuring the correct placement of '+' and '-' signs
        result = ''.join([terms[0]] + [f"+{term}" if not term.startswith('-') else term for term in terms[1:]])

        return result
